fairseq/tasks/fairseq_task.py

```diff
@@ -138,9 +138,17 @@ def get_batch_iterator(
             )
 
         # create mini-batches with given size constraints
-        batch_sampler = data_utils.batch_by_size_tpu(
-            indices, dataset.num_tokens, self.args.input_shapes
-        )
+        if getattr(self.args, 'use_gpu', True):
+            batch_sampler = data_utils.batch_by_size(
+                indices, dataset.num_tokens, max_tokens=max_tokens,
+                max_sentences=max_sentences,
+                required_batch_size_multiple=required_batch_size_multiple,
+            )
+        else:
+            batch_sampler = data_utils.batch_by_size_tpu(
```

Comment:
Where is this implemented?
Did I review it?

Reply:
yeah, it was implemented a long time ago, let me find the pr.

Reply:
This needs to be "world aware" otherwise you will end up with issues at the end of the samples set.

Reply:
so, [this is the pr](https://github.com/pytorch-tpu/fairseq/commit/7a9e90e66b095692ace9d583fd421b745308e014#diff-47dbda9fcad726ba8d5aa806c9af04ee) that made the `tpu` branch in this repo. But it was really first implemented in our `examples` repo, when our runner was still there. See [here](https://github.com/pytorch-tpu/examples/blob/xrt.r1.15/fairseq_train_tpu.py#L75)

Reply:
when we moved our runner inside the `fairseq` repo, it got carried over.

Reply:
ah here is the [actual pr](https://github.com/pytorch-tpu/examples/pull/11) that introduced `batch_by_size_tpu`

Reply:
The classes that uses this sampler downstream are world aware. Unfortunately, due to the `.ceil` thing discussed before, we do get empty batches at the end for some cores. We definitely do always get the same number of iterations per core. All we then have to do is to drop the last batches, which we do [here](https://github.com/pytorch-tpu/fairseq/blob/tpu/train.py#L361)